### DISCUSSION TITLE
Environment variables not working

### DISCUSSION THREAD

**Discussion by lkdevacc** [vercel/next.js#28469] [Help]
Next version is `11.1.0`
It is worth mentioning that I installed Next with tailwind using `npx create-next-app -e with-tailwindcss my-project`
I created a` .env.local` file in the root of the project with the following line: `API_URL="http://localhost:8088/api"
`
And when I try to call it from the code it seems like it is not interpreting the string literal.
this is the code:

```
export async function getServerSideProps() {  
  const res = await fetch(`${process.env.API_URL}/products/`, {
    method: 'GET',
  })
  const { data } = await res.json()

  return { props: { data } };
}
```

and this is the line in the log on the browser that shows the error:

```
TypeError: Only absolute URLs are supported
> 45 | const res = await fetch(`${process.env.API_URL}/items/`, {
```

**Comment by cryskram** [comment]
make it `API_URL=http://localhost:8088/api`

And by the way, why are u storing the URL in a `.env` file

**Reply by lkdevacc** [reply]
I want the API to be hidden / secret. I also realized by that time that it's not possible to achieve it that way because I'm trying to get it on the client side so I always get undefined. What is the correct way then to store external API URL and make it secret on the `getServerSideProps`?

**Reply by cryskram** [reply]
U can have the API URL in the `getServerSideProps` only since no one will be able to see the URL, (works on the server) unless u make the project open source. This is the only way out... though anyone can go to the Network tab in the Dev Tools and check from where the operations are taking place

**Reply by lkdevacc** [reply]
Thanks, my mistake was that I did not build my project before starting so I did not see the environment variables. But after your first reply I have a new question: Why did you ask why I'm storing it in a `.env` file? Is it not a good way to store secret stuff? Or you asked why I think an API URL should be secret - should I not make it hidden?

**Comment by cryskram** [comment]
oh.. good to see that ur work is working now... Just asked because I have never stored the API URL in a `.env` file. You can and u should store secret variables in a even file....

The point is that though u make the API URL hidden or not, people can goto the network tab in the developer tools and check from where the data is sent and received... so practically no use of make the URL private since it is indirectly public to everyone. This is want I meant to say

**Reply by lkdevacc** [reply]
Good to know, so I will just store it in some public file. Where should it be - a `.env` file with `NEXT_PUBLIC_` prefix, or maybe something like `publicRuntimeConfig` in the `next.config.js` file as seen in the docs:

```
module.exports = {
    publicRuntimeConfig: {
      // Will be available on both server and client
      publicVariable: 'public',
    },
  }
```

**Reply by cryskram** [reply]
Yea u can do that

**Reply by lkdevacc** [reply]
Which one is the most preferable way (And why?)

**Reply by cryskram** [reply]
both of them are fine. `env` file usage has been for a long time while using `next.config.js` is only for Next.js projects. I haven't used the next config file much, but have seen them being used. 
I personally would prefer to use the env file in general since at the time of hosting a project u need an env file or need to add it manually

On the other hand, next.config.js makes this easy.... Here's what I do for my nextjs projects.
Add variables into the `.env` file then will call `dotenv` package inside the `next.config.js` file and use and provide the variables there. In this way, u need not add the variables manually at the time of hosting. Everything will be taken care by the config file

Both have pros and cons as mentioned above... the choice is yours
